### ISSUE TITLE
Freez parsing sql statement with many brackets

### ISSUE THREAD

**Issue by stefardin** [dbeaver/dbeaver#14735]
#### System information: 
- Windows 10 Pro for Workstations 21H1
- DBeaver version 21.3.0.202111281534
- Additional extensions: no one

#### Connection specification:
- with both Oracle or PostgreSQL local or remote database

#### Describe the problem you're observing:
SQL statements with many brackets causes freezing before start execution. It depends on many brackets. Same statement with less brackets work fine. 

#### Steps to reproduce, if exist:
Statement in attachment
[statement.txt](https://github.com/dbeaver/dbeaver/files/7625434/statement.txt)

#### Include any warning/errors/backtraces from the logs
no error or log found.

**Comment by LonwoLonwo** [comment]
Hello @stefardin 
Thanks for the bug report.

Close cases: https://github.com/dbeaver/dbeaver/issues/14053 and https://github.com/dbeaver/dbeaver/issues/13905

Maybe we must update the @JSQLParser library in the project.

**Comment by manticore-projects** [comment]
Greetings!

Definitely related to JSQLParser, although I have not been able to spot the problem immediately.
I will look into it and return with a solution shortly.

**Comment by manticore-projects** [comment]
The problem should be related to the `IN` clause which provide just too many potential `Complex Expressions` which are expensive to parse.

I will need to modify JSQLParser in this regard (limiting `Complex Expressions` to max. 10, switching to cheap simple expressions beyond that).

**Comment by manticore-projects** [comment]
As it looks to me, your SQL is invalid since the opening/closing brackets do not match.

**Comment by manticore-projects** [comment]
After a bit of trial and error and fixing the query by removing the last 2 closing brackets:

1) the query parses, just (too) slow (e. g. within 3 seconds on my computer)

2) the reason is the Bracket-Structure, because everything inside those Brackets is seen as an expensive Complex Expression as long as we do not exceed a Nesting Depth of 10.

3) Simple Expression (only) parsing would be applied automatically when Nesting Depth exceeded 10.  However, we can't apply Simple Expression parsing because of the `IN ( SELECT ... )` clause which again is a Complex Expression. 

To cut the story short: This particular ill-written query brings JSQLParser to its limits and I am not sure if I will be able to find a work around. (I will try though.) You will need to accept, that JSQLParser based on its nature as a generic, RDBMS agnostic parser is sensitive to nested complex expressions.

A complete solution would consist of 2 steps:

1) rewrite the query and avoid those redundant brackets, as they do not provide any added value to the query

2) wrap any call to JSQLParser into a SwingWorker and execute it with a timeout (of 2 seconds or so) in order to prevent freeing when a query simply takes too long.

I still will try to work around this particular challenge.

**Comment by LonwoLonwo** [comment]
> wrap any call to JSQLParser into a SwingWorker and execute it with a timeout (of 2 seconds or so) in order to prevent freeing when a query simply takes too long.

Yep, SwingWorker is cool, but we just do not use Swing in DBeaver.

Thanks for the answer @manticore-projects

**Comment by manticore-projects** [comment]
I have opened [JSQLParser Issue 1438](https://github.com/JSQLParser/JSqlParser/issues/1438) for this.

**Comment by manticore-projects** [comment]
> Yep, SwingWorker is cool, but we just do not use Swing in DBeaver.

For Eclipse RCP, you should use Jobs API:
```java
// get UISynchronize injected as field
@Inject UISynchronize sync;

Job job = Job.create("Parse Statements", (ICoreRunnable) monitor -> {
    // do something long running
    // e. g. Parse the SQL Statement

    // If you want to update the UI
    sync.asyncExec(() -> {
        // do something in the user interface
        // e.g. Print the formatted SQL Statement
    });
});

// Start the Job
job.schedule();
```

**Comment by manticore-projects** [comment]
Solution provided by [PR 1439](https://github.com/JSQLParser/JSqlParser/pull/1439), pending approval.
I have added a 6 seconds timeout to the Parser, which will fail gracefully when it is not able to process the SQL Statement within that time.

This should stop DBeaver from freezing on wrong statements effectively:
1) the corrected SQL Statement provided above will parser within 1 second
2) the original, syntactically wrong SQL Statement provided above will timeout gracefully within 6 seconds

You may want to use the Snapshot below as bridge-gap solution until the PR is accepted (which may take a while):
[jsqlparser-4.3-SNAPSHOT.zip](https://github.com/dbeaver/dbeaver/files/7696822/jsqlparser-4.3-SNAPSHOT.zip)

Please do let me know if anything was left open.

**Comment by uslss** [comment]
fixed by #15155
